src/redfish.c

```diff
@@ -0,0 +1,891 @@
+            }
+            char type_instance[sizeof(v1.type_instance)];
+            int ch_count = snprintf(type_instance, sizeof(type_instance), "%d",
+                                    (int)json_integer_value(member_id));
+            if (ch_count != 0)
+              sstrncpy(v1.type_instance, type_instance,
+                       sizeof(v1.type_instance));
+            else {
+              ERROR("Failed to convert MemberId to a character");
+              continue;
+            }
+          }
+          /* Checking whether real or integer value */
+          redfish_value_t value;
+          redfish_value_type_t type = VAL_TYPE_STR;
+          if (json_is_string(object)) {
+            value.string = (char *)json_string_value(object);
+          } else if (json_is_integer(object)) {
+            type = VAL_TYPE_INT;
+            value.integer = json_integer_value(object);
+          } else if (json_is_real(object)) {
+            type = VAL_TYPE_REAL;
+            value.real = json_real_value(object);
+          }
+          const data_set_t *ds = plugin_get_ds(prop->type);
+
+          /* Check if data set found */
+          if (ds == NULL)
+            continue;
+
+          v1.values = &(value_t){0};
+          redfish_convert_val(&value, type, v1.values, ds->ds[0].type);
+
+          sstrncpy(v1.host, serv->host, sizeof(v1.host));
+          sstrncpy(v1.plugin, PLUGIN_NAME, sizeof(v1.plugin));
+          sstrncpy(v1.type, prop->type, sizeof(v1.type));
+          plugin_dispatch_values(&v1);
+          /* Clear values assigned in case of leakage */
+          v1.values = NULL;
+          v1.values_len = 0;
+        }
+      }
+      json_decref(json_array);
+    }
+  } else {
+    WARNING("Failed to get payload for service name \"%s\"", serv->name);
+  }
+}
+
+static int redfish_read(__attribute__((unused)) user_data_t *ud) {
+  for (llentry_t *le = llist_head(ctx->services); le != NULL; le = le->next) {
+    redfish_service_t *service = (redfish_service_t *)le->value;
+    for (llentry_t *le = llist_head(service->query_ptrs); le != NULL;
+         le = le->next) {
+      redfish_query_t *query = (redfish_query_t *)le->value;
+      redfish_payload_ctx_t rs = {.service = service,
+                                  .resources = query->resources};
+      getPayloadByPathAsync(service->redfish, query->endpoint, NULL,
+                            redfish_process_payload, &rs);
+      /* TODO: Work around for race condition. Needs permanent fix. */
+      sleep(10);
+      // serviceDecRefAndWait(service->redfish);
+    }
+  }
+  return 0;
+}
+
+static int redfish_cleanup(void) {
+  for (llentry_t *le = llist_head(ctx->services); le; le = le->next) {
+    redfish_service_t *service = (redfish_service_t *)le->value;
+
+    cleanupServiceEnumerator(service->redfish);
+    for (size_t i = 0; i < service->queries_num; i++)
+      sfree(service->queries[i]);
+
+    llist_destroy(service->query_ptrs);
+
+    sfree(service->name);
```

Comment:
I've seen this list of sfree statements for various service members before. Maybe put that into a `service_destroy()` function?